### PR TITLE
feat(nms): add rate limiting on a subset of endpoints

### PR DESCRIPTION
**Pull request by pierre-roussel**
## Summary

Adds rate limiting to a small subset of computationally expensive endpoints in NMS, outlined in https://github.com/magma/magma/issues/15348:
- [ ] `/user/login`
- [ ] `/user/logout`
- [ ] `/login/oidc/callback`
- [ ] `/user/login/oidc/callback`

## Test Plan

Middleware is unit tested to ensure proper behavior, local testing was used to confirm the desired behavior.

On rate limited endpoints:
```
curl localhost:8081/user/login?[1-100]
```

After 100 request within 15 minutes for a given IP, a 429: too many request error code is returned by the endpoint for the offending IP, other endpoints still work as expected.

Other rate limited endpoints to test:
- [ ] `curl localhost:8081/user/logout`
- [ ] `curl localhost:8081/login/oidc/callback`
- [ ] `curl localhost:8081/user/login/oidc/callback`

## Additional Information

The configuration for the rate limit is located under: `magma/nms/config/config.ts` in `RATE_LIMIT_CONFIG`.

To be determined: 
- the exact configuration method, a simple node.js `const` is used at the moment, switching to an environment variable would allow for per deploy configuration
- the base rate limiting config is set to 100 request within a 15min time window (this seems fairly conservative but I've got limited context on those endpoints)

### CODE REVIEW

File: nms/config/config.ts [hunk 1]
```diff
@@ -13,12 +13,19 @@
 
 import fs from 'fs';
 import {getValidLogLevel} from '../shared/logging';
+import {Options as RateLimitOptions} from 'express-rate-limit';
```
[the remark this answers is not in the document]
Comment (pierre-roussel) [reply nms/config/config.ts]: Good to know I can't trust my eslint locally 😅 
```
yarn run eslint
yarn run v1.22.21
$ ./node_modules/.bin/eslint --ignore-path .eslintignore
Done in 5.88s.
```

**Comment by pierre-roussel** [comment]
ℹ️ Just added a PGP key and signed all commits

**Comment by lucasgonze** [comment]
I don't think the failing test could be caused by your changes.

Requests:
- Implement the environment variable approach to augment the const. (We'll also have to figure out how to document it, but that can be a separate project).
- Rebase against current master.

File: nms/config/config.ts [hunk 2]
```diff
@@ -12,13 +12,20 @@
...
 export const DEV_MODE = process.env.NODE_ENV !== 'production';
 export const LOG_FORMAT = DEV_MODE ? 'shell' : 'json';
 export const LOG_LEVEL = getValidLogLevel(process.env.LOG_LEVEL);
 export const LOGGER_HOST = process.env.LOGGER_HOST || 'fluentd:9880';
 export const API_HOST = process.env.API_HOST || 'magma_test.local';
+export const RATE_LIMIT_CONFIG: Partial<RateLimitOptions> = {
+  windowMs: Number(process.env.RATE_LIMIT_WINDOW) || 15 * 60 * 1000, // 15 min default
+  limit: Number(process.env.RATE_LIMIT_COUNT) || 100,
+  standardHeaders: 'draft-7',
+  legacyHeaders: false,
```
Comment (jahid-wavelabs) [inline nms/config/config.ts:28]: A default message is expected to be included.

**Comment by lucasgonze** [comment]
Consensus on accepting default rate given that environment variable exists as a failsafe.

Need to add documentation on the rate and variable. Lucas to start a Discussions thread.

File: nms/config/config.ts (hunk 2 above)
Comment (pierre-roussel) [reply nms/config/config.ts:28]: The default is "Too many requests, please try again later.", what do you recommend?

Comment (jahid-wavelabs) [reply nms/config/config.ts:28]: Please go ahead and add the above message provided by you to the default.

**Comment by pierre-roussel** [comment]
I don't have enough context on the endpoints to decide on anything but I'll share my thought process for the initial values here: in a normal workflow I usually don't expect a single IP to hit an OIDC login/callbacks/logout endpoint more than once or twice per token expiration so a 100 request/15min window seemed generous enough at a glance (mind you, I don't know the chosen expiration time of those tokens). 

It should be trivial to make rate limiting less aggressive if the load on servers is still too high after a deploy but it becomes much harder to justify doing the opposite and risk locking people out of the management UI. After a quick chat with @tmdzk it seems that the some of the endpoints are used by clients sharing the same IP so decreasing the window and limit could make more sense (10 request/1s window for example).

Please share any thoughts on that and feel free to update the value as needed in the config 👍

**Comment by lucasgonze** [comment]
Posted a general notice of the architectural change to Slack.

**Comment by lucasgonze** [comment]
Nobody responded to request for comments within the 48 hour window, so we are good to go on community feedback.

**Comment by lucasgonze** [comment]
I have approved. Aside from functional testing, the only remaining work is documenting the environment variable.

**Comment by lucasgonze** [comment]
I initialized a documentation page at https://github.com/magma/magma/wiki/Rate-Limiting-on-NMS

File: .github/workflows/nms-workflow.yml [hunk 3]
```diff
@@ -146,7 +146,7 @@ jobs:
       - name: run yarn test
         run: |
           cd ${MAGMA_ROOT}/nms
-          yarn add jest@^26.4.2 -W --dev
+          yarn add jest@^28.1.3 -W --dev
```
Comment (pierre-roussel) [inline .github/workflows/nms-workflow.yml:149]: There's a version mismatch of jest in the CI pipeline, I'm not sure if this is causing the test errors but `jest.useFakeTimers();` is valid code and should be supported as it's required for unit testing the rate limiter. 

Jest 26 should support this as well and the error I'm seeing is: https://github.com/magma/magma/actions/runs/8312668614/job/22990589385
```
 Summary of all failing tests
FAIL server/middleware/__tests__/rateLimit-test.ts
  ● Test suite failed to run

    TypeError: Cannot assign to read only property 'performance' of object '[object global]'
```
I'll attempt to use Jest 29 in the CI if this still happens as it seems to have helped in this issue https://github.com/jestjs/jest/issues/14741

Comment (lucasgonze) [reply .github/workflows/nms-workflow.yml:149]: We discussed the jest upgrade in the last TSC and had consensus to go with it.

Comment (pierre-roussel) [reply .github/workflows/nms-workflow.yml:149]: It does not seem to have worked, I'm not sure I fully get why the CI prevents us from replacing the timer implementation at the moment. If anyone has an idea as to why that might be the case here feel free to chime in, here's the error:
```
Summary of all failing tests
FAIL server/middleware/__tests__/rateLimit-test.ts
  ● Test suite failed to run

    TypeError: Cannot assign to read only property 'performance' of object '[object global]'

      27 | };
      28 |
    > 29 | jest.useFakeTimers();
         |      ^
      30 | jest.mock('../../../config/config', () => config);
      31 |
      32 | describe('Rate limit test', () => {

      at hijackMethod (node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:946:32)
      at Object.install (node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:1733:17)
      at FakeTimers.useFakeTimers (node_modules/@jest/fake-timers/build/modernFakeTimers.js:110:36)
      at Object.useFakeTimers (server/middleware/__tests__/rateLimit-test.ts:29:6)


Test Suites: 1 failed, 4 skipped, 53 passed, 54 of 58 total
Tests:       4 skipped, 351 passed, 355 total
Snapshots:   0 total
Time:        133.391 s
Ran all test suites in 2 projects.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

File: nms/server/middleware/__tests__/rateLimit-test.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+const config = {
+  RATE_LIMIT_CONFIG: {
+    windowMs: 15 * 1000, // 15 seconds
+    limit: 2,
+  },
+};
+
+Object.defineProperty(global, "performance", {
+  writable: true,
+});
```
Comment (pierre-roussel) [inline nms/server/middleware/__tests__/rateLimit-test.ts]: Attempting this, pulled from https://github.com/facebook/react-native/issues/35701#issuecomment-1847579429

**Comment by pierre-roussel** [comment]
Closing as this was merged by @tmdzk here: https://github.com/magma/magma/commit/2d2fb7049f569bd58510eefeff523526e39ed4cc

**Pull request closed without merging**